packages/drivers/odsp-driver/src/odspSnapshotParser.ts

```diff
@@ -0,0 +1,74 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, stringToBuffer } from "@fluidframework/common-utils";
+import * as api from "@fluidframework/protocol-definitions";
+import { IOdspSnapshot, IOdspSnapshotCommit } from "./contracts";
+import { ISnapshotValue } from "./odspUtils";
+
+/**
+ * Build a tree hierarchy base on a flat tree
+ *
+ * @param flatTree - a flat tree
+ * @param blobsShaToPathCache - Map with blobs sha as keys and values as path of the blob.
+ * @returns the hierarchical tree
+ */
+function buildHierarchy(flatTree: IOdspSnapshotCommit): api.ISnapshotTree {
+    const lookup: { [path: string]: api.ISnapshotTree } = {};
+    // id is required for root tree as it will be used to determine the version we loaded from.
+    const root: api.ISnapshotTree = { id: flatTree.id, blobs: {}, commits: {}, trees: {} };
+    lookup[""] = root;
+
+    for (const entry of flatTree.entries) {
+        const lastIndex = entry.path.lastIndexOf("/");
+        const entryPathDir = entry.path.slice(0, Math.max(0, lastIndex));
+        const entryPathBase = entry.path.slice(lastIndex + 1);
+
+        // ODSP snapshots are created breadth-first so we can assume we see tree nodes prior to their contents
+        const node = lookup[entryPathDir];
+
+        // Add in either the blob or tree
+        if (entry.type === "tree") {
+            const newTree: api.ISnapshotTree = {
+                blobs: {},
+                commits: {},
+                trees: {},
+                unreferenced: entry.unreferenced,
+            };
+            node.trees[decodeURIComponent(entryPathBase)] = newTree;
+            lookup[entry.path] = newTree;
+        } else if (entry.type === "blob") {
+            node.blobs[decodeURIComponent(entryPathBase)] = entry.id;
+        } else if (entry.type === "commit") {
+            node.commits[decodeURIComponent(entryPathBase)] = entry.id;
+        }
+    }
+
+    return root;
+}
+
+/**
+ * Converts existing IOdspSnapshot to snapshot tree, blob array and ops
+ * @param odspSnapshot - snapshot
+ */
+export function convertOdspSnapshotToSnapsohtTreeAndBlobs(
+    odspSnapshot: IOdspSnapshot,
+): ISnapshotValue {
+    const blobsWithBufferContent = new Map<string, ArrayBuffer>();
+    if (odspSnapshot.blobs) {
+        odspSnapshot.blobs.forEach((blob) => {
+            assert(blob.encoding === "base64" || blob.encoding === undefined,
+                0x0a4 /* `Unexpected blob encoding type: '${blob.encoding}'` */);
+            blobsWithBufferContent.set(blob.id, stringToBuffer(blob.content, blob.encoding ?? "utf8"));
+        });
+    }
+    const val: ISnapshotValue = {
+        blobs: blobsWithBufferContent,
+        ops: odspSnapshot.ops ?? [],
+        sequenceNumber: odspSnapshot.trees && (odspSnapshot.trees[0] as any).sequenceNumber,
```

Comment:
It would be great to assert here that odspSnapshot.trees.length === 1.
I assume that relies on work Marcelo did recently to remove commits from response. If so, would be great to confirm it is enabled in prod and for all generations of files.
Also can you please rename IOdspSnapshot.trees to IOdspSnapshot.commits? I made it's type to be IOdspSnapshotCommit to better reflect meaning, but never renamed actual variables...

Reply:
Yes, I confirmed him before, he said by  mid-end of june. I will confirm once more.

Reply:
Renamed trees to commits.

Reply:
I think we need to wait a bit more for adding that assert. With that change, we would also be able to remove some code where we used to handle that.

Reply:
Actually we cannot rename trees to commits. It is the type that we fetch from server and we get trees from server. Also I think with new format we are calling trees and IOdspSnapshot would be totally gone. So we can leave it as such.

Reply:
You are right about rename :)
If we are not asserting and getting two entries there, what does it mean? Don't we have .protocol commit in first entry, and simply ignore .app enrty?